metals/src/main/scala/scala/meta/internal/metals/BuildTargets.scala

```diff
@@ -642,4 +511,310 @@ object BuildTargets {
       leaves: collection.Set[BuildTargetIdentifier]
   )
 
+  trait Data {
+    def sourceItemsToBuildTarget
+        : scala.collection.Map[AbsolutePath, ConcurrentLinkedQueue[
+          BuildTargetIdentifier
+        ]]
+    def buildTargetInfo
+        : scala.collection.Map[BuildTargetIdentifier, BuildTarget]
+    def javaTargetInfo: scala.collection.Map[BuildTargetIdentifier, JavaTarget]
+    def scalaTargetInfo
+        : scala.collection.Map[BuildTargetIdentifier, ScalaTarget]
+    def inverseDependencies
+        : scala.collection.Map[BuildTargetIdentifier, ListBuffer[
+          BuildTargetIdentifier
+        ]]
+    def buildTargetSources
+        : scala.collection.Map[BuildTargetIdentifier, util.Set[AbsolutePath]]
+    def inverseDependencySources
+        : scala.collection.Map[AbsolutePath, Set[BuildTargetIdentifier]]
+    def sourceJarNameToJarFile: scala.collection.Map[String, AbsolutePath]
+    def isSourceRoot: java.util.Set[AbsolutePath]
+    // if workspace contains symlinks, original source items are kept here and source items dealiased
+    def originalSourceItems: java.util.Set[AbsolutePath]
+
+    def targetToConnection
+        : scala.collection.Map[BuildTargetIdentifier, BuildServerConnection]
+    def sourceBuildTargets(
+        sourceItem: AbsolutePath
+    ): Option[Iterable[BuildTargetIdentifier]]
+
+    def allTargetRoots: Iterator[AbsolutePath]
+    def all: Iterator[BuildTarget]
+    def allBuildTargetIds: Seq[BuildTargetIdentifier]
+    def allScala: Iterator[ScalaTarget]
+    def allJava: Iterator[JavaTarget]
+    def scalaTarget(id: BuildTargetIdentifier): Option[ScalaTarget]
+    def javaTarget(id: BuildTargetIdentifier): Option[JavaTarget]
+
+    def targetRoots(buildTarget: BuildTargetIdentifier): List[AbsolutePath]
+    def javaTargetRoot(buildTarget: BuildTargetIdentifier): Option[AbsolutePath]
+    def scalaTargetRoot(
+        buildTarget: BuildTargetIdentifier
+    ): Option[AbsolutePath]
+
+    def info(id: BuildTargetIdentifier): Option[BuildTarget]
+    def targetJarClasspath(
+        id: BuildTargetIdentifier
+    ): Option[List[AbsolutePath]]
+    def targetClasspath(id: BuildTargetIdentifier): Option[List[String]]
+    def targetClassDirectories(id: BuildTargetIdentifier): List[String]
+
+    def allWorkspaceJars: Iterator[AbsolutePath]
+
+    def addSourceItem(
+        sourceItem: AbsolutePath,
+        buildTarget: BuildTargetIdentifier
+    ): Unit
+  }
+
+  final case class DataSeq(head: WritableData, tail: List[WritableData]) {
```

Comment:
Maybe a silly question, but I'm assuming this is mainly for the utilities that you have in here like `fromIterators` and `fromOptions`, but why not just have `DataSeq` be a wrapper around `List[WritableData]`? Why have a head and the tail here?

Reply:
Seems I wanted to ensure it's non-empty, although I can't recall why… and it seems it's not needed.

Reply:
Ok, changed that to `final case class DataSeq(list: List[TargetData])`